File: src/main/java/org/blitzortung/android/alert/handler/AlertStatusHandler.java
package org.blitzortung.android.alert.handler;

import android.location.Location;
import android.util.Log;
import org.blitzortung.android.alert.AlertParameters;
import org.blitzortung.android.alert.AlertResult;
import org.blitzortung.android.alert.object.AlertSector;
import org.blitzortung.android.alert.object.AlertStatus;
import org.blitzortung.android.app.Main;
import org.blitzortung.android.data.beans.Strike;

import java.util.Collection;
import java.util.Locale;
import java.util.SortedMap;
import java.util.TreeMap;

public class AlertStatusHandler {

    private final AlertSectorHandler alertSectorHandler;
    private final AlertParameters alertParameters;

    public AlertStatusHandler(AlertSectorHandler alertSectorHandler, AlertParameters alertParameters) {
        this.alertSectorHandler = alertSectorHandler;
        this.alertParameters = alertParameters;
    }

    public AlertStatus checkStrikes(AlertStatus alertStatus, Collection<? extends Strike> strikes, Location location) {

        alertStatus.clearResults();

        long thresholdTime = System.currentTimeMillis() - alertParameters.getAlarmInterval();

        alertSectorHandler.setCheckStrikeParameters(location, thresholdTime);

        Location strikeLocation = new Location("");

        for (Strike strike : strikes) {
            float bearingToStrike = location.bearingTo(strike.getLocation(strikeLocation));

            AlertSector alertSector = getSectorForBearing(alertStatus, bearingToStrike);
            alertSectorHandler.checkStrike(alertSector, strike);
        }
        return alertStatus;
    }


    public long getLatestTimstampWithin(float distanceLimit, AlertStatus alertStatus) {
        long latestTimestamp = 0;
        
        for (AlertSector sector : alertStatus.getSectors()) {
            latestTimestamp = Math.max(latestTimestamp, alertSectorHandler.getLatestTimestampWithin(distanceLimit, sector));
        }
        
        return latestTimestamp;
    }
    
    public AlertSector getSectorWithClosestStrike(AlertStatus alertStatus) {
        float minDistance = Float.POSITIVE_INFINITY;

        AlertSector sectorWithClosestStrike = null;
        for (AlertSector sector : alertStatus.getSectors()) {
            if (sector.getClosestStrikeDistance() < minDistance) {
                minDistance = sector.getClosestStrikeDistance();
                sectorWithClosestStrike = sector;
            }
        }
        return sectorWithClosestStrike;
    }

    public AlertResult getCurrentActivity(AlertStatus alertStatus) {
        AlertSector sector = getSectorWithClosestStrike(alertStatus);

        return sector != null ? new AlertResult(sector, alertParameters.getMeasurementSystem().getUnitName()) : null;
    }

    public String getTextMessage(AlertStatus alertStatus, float notificationDistanceLimit) {
        SortedMap<Float, AlertSector> distanceSectors = getSectorsSortedByClosestStrikeDistance(alertStatus, notificationDistanceLimit);

        StringBuilder sb = new StringBuilder();

        if (distanceSectors.size() > 0) {
            for (AlertSector sector : distanceSectors.values()) {
                sb.append(sector.getLabel());
                sb.append(" ");
                sb.append(String.format("%.0f%s", sector.getClosestStrikeDistance(), alertParameters.getMeasurementSystem().getUnitName()));
                sb.append(", ");
            }
            sb.setLength(sb.length() - 2);
        }

        return sb.toString();
    }

    private SortedMap<Float, AlertSector> getSectorsSortedByClosestStrikeDistance(AlertStatus alertStatus, float notificationDistanceLimit) {
        SortedMap<Float, AlertSector> distanceSectors = new TreeMap<Float, AlertSector>();

        for (AlertSector sector : alertStatus.getSectors()) {
            if (sector.getClosestStrikeDistance() <= notificationDistanceLimit) {
                distanceSectors.put(sector.getClosestStrikeDistance(), sector);
            }
        }
        return distanceSectors;
    }

    private AlertSector getSectorForBearing(AlertStatus alertStatus, double bearing) {
        for (AlertSector sector : alertStatus.getSectors()) {
            if (sectorContainsBearing(sector, bearing)) {
                return sector;
            }
        }
        Log.w(Main.LOG_TAG, String.format(Locale.ENGLISH, "AlarmStatusHandler.getSectorForBearing(): no sector for bearing %.2f found", bearing));
        return null;
    }

    private boolean sectorContainsBearing(AlertSector sector, double bearing) {
        float minimumSectorBearing = sector.getMinimumSectorBearing();
        float maximumSectorBearing = sector.getMaximumSectorBearing();

        if (maximumSectorBearing > minimumSectorBearing) {
            return bearing < maximumSectorBearing && bearing >= minimumSectorBearing;
        } else {
            return bearing >= minimumSectorBearing || bearing < maximumSectorBearing;
        }
    }
}

File: app/src/main/java/org/blitzortung/android/alert/handler/AlertStatusHandler.kt
package org.blitzortung.android.alert.handler

import android.location.Location
import android.util.Log

import org.blitzortung.android.alert.AlertResult
import org.blitzortung.android.alert.data.AlertSector
import org.blitzortung.android.alert.data.AlertContext
import org.blitzortung.android.app.Main
import org.blitzortung.android.data.beans.Strike
import java.util.Locale
import java.util.SortedMap
import java.util.TreeMap

class AlertStatusHandler(private val alertSectorHandler: AlertSectorHandler) {

    fun checkStrikes(alertContext: AlertContext, strikes: Collection<Strike>, location: Location): AlertContext {

        val thresholdTime = System.currentTimeMillis() - alertContext.alertParameters.alarmInterval

        alertSectorHandler.setCheckStrikeParameters(location, thresholdTime)

        val strikeLocation = Location("")

        for (strike in strikes) {
            val bearingToStrike = calculateBearingToStrike(location, strikeLocation, strike)

            val alertSector = getRelevantSector(bearingToStrike.toDouble(), alertContext)
            alertSector?.let { alertSectorHandler.checkStrike(alertSector, strike, alertContext) }
        }
        return alertContext
    }

    private fun calculateBearingToStrike(location: Location, strikeLocation: Location, strike: Strike): Float {
        strikeLocation.longitude = strike.longitude.toDouble()
        strikeLocation.latitude = strike.latitude.toDouble()
        return location.bearingTo(strikeLocation)
    }


    fun getLatestTimstampWithin(distanceLimit: Float, alertContext: AlertContext): Long {
        return alertContext.sectors.fold(0L, {
            latestTimestamp, sector ->
            Math.max(latestTimestamp, alertSectorHandler.getLatestTimestampWithin(distanceLimit, sector))
        })
    }

    fun getSectorWithClosestStrike(alertContext: AlertContext): AlertSector? {
        return alertContext.sectors
                .filter { it.closestStrikeDistance < Float.POSITIVE_INFINITY }
                .sortedBy { it.closestStrikeDistance }
                .firstOrNull()
    }

    fun getCurrentActivity(alertContext: AlertContext?): AlertResult? {
        if (alertContext != null) {
            val sector = getSectorWithClosestStrike(alertContext)

            return if (sector != null) AlertResult(sector, alertContext.alertParameters.measurementSystem.unitName) else null
        }
        return null
    }

    fun getTextMessage(alertContext: AlertContext, notificationDistanceLimit: Float): String {
        val distanceSectors = getSectorsSortedByClosestStrikeDistance(alertContext, notificationDistanceLimit)

        val sb = StringBuilder()

        if (distanceSectors.size > 0) {
            for (sector in distanceSectors.values) {
                sb.append(sector.label)
                sb.append(" ")
                sb.append("%.0f%s".format(sector.closestStrikeDistance, alertContext.alertParameters.measurementSystem.unitName))
                sb.append(", ")
            }
            sb.setLength(sb.length - 2)
        }

        return sb.toString()
    }

    private fun getSectorsSortedByClosestStrikeDistance(alertContext: AlertContext, notificationDistanceLimit: Float): Map<Float, AlertSector> {
        return alertContext.sectors
                .filter { it.closestStrikeDistance <= notificationDistanceLimit }
                .sortedBy { it.closestStrikeDistance }
                .toMapBy { it.closestStrikeDistance }
    }

    private fun getRelevantSector(bearing: Double, alertContext: AlertContext): AlertSector? {
        return alertContext.sectors.firstOrNull {sectorContainsBearing(it, bearing) }
    }

    private fun sectorContainsBearing(sector: AlertSector, bearing: Double): Boolean {
        val minimumSectorBearing = sector.minimumSectorBearing
        val maximumSectorBearing = sector.maximumSectorBearing

        if (maximumSectorBearing > minimumSectorBearing) {
            return bearing < maximumSectorBearing && bearing >= minimumSectorBearing
        } else {
            return bearing >= minimumSectorBearing || bearing < maximumSectorBearing
        }
    }
}
